src/software/field_tests/field_test_fixture.py

```diff
@@ -0,0 +1,435 @@
+            raise Exception("robot positioning not set")
+
+    def time_provider(self):
+        """Provide the current time in seconds since the epoch"""
+
+        with self.timestamp_mutex:
+            return self.timestamp
+
+    def run_test(
+        self,
+        always_validation_sequence_set=[[]],
+        eventually_validation_sequence_set=[[]],
+        test_timeout_s=3,
+    ):
+        """Run a test. In a field test this means beginning validation.
+
+        :param always_validation_sequence_set: Validation functions that should
+                                hold on every tick
+        :param eventually_validation_sequence_set: Validation that should
+                                eventually be true, before the test ends
+        :param test_timeout_s: The timeout for the test, if any eventually_validations
+                                remain after the timeout, the test fails.
+        """
+
+        test_end_time = time.time() + test_timeout_s
+
+        while time.time() < test_end_time:
+
+            # Update the timestamp logged by the ProtoLogger
+            with self.timestamp_mutex:
+                ssl_wrapper = self.ssl_wrapper_buffer.get(block=False)
+                self.timestamp = ssl_wrapper.detection.t_capture
+
+            while True:
+                try:
+                    world = self.world_buffer.get(
+                        block=True, timeout=WORLD_BUFFER_TIMEOUT
+                    )
+                    break
+                except queue.Empty as empty:
+                    # If we timeout, that means full_system missed the last
+                    # wrapper and robot status, lets resend it.
+                    logger.warning("No world received")
+
+            # Validate
+            (
+                eventually_validation_proto_set,
+                always_validation_proto_set,
+            ) = validation.run_validation_sequence_sets(
+                world,
+                eventually_validation_sequence_set,
+                always_validation_sequence_set,
+            )
+
+            if self.publish_validation_protos:
+                # Set the test name
+                eventually_validation_proto_set.test_name = self.test_name
+                always_validation_proto_set.test_name = self.test_name
+
+                # Send out the validation proto to thunderscope
+                self.blue_full_system_proto_unix_io.send_proto(
+                    ValidationProtoSet, eventually_validation_proto_set
+                )
+                self.blue_full_system_proto_unix_io.send_proto(
+                    ValidationProtoSet, always_validation_proto_set
+                )
+
+            # Check that all always validations are always valid
+            validation.check_validation(always_validation_proto_set)
+
+        # Check that all eventually validations are eventually valid
+        validation.check_validation(eventually_validation_proto_set)
+
+
+def field_test_initializer(
```

Comment:
nit: I wonder if there is a clean way of refactoring this function and `simulated_test_initializer`, since other than the initialization of `Simulator`/`RobotCommunication` and the test runner, the two functions seem to be identical.